advanced/button-interrupt/examples/solution.rs

```diff
@@ -1,70 +1,36 @@
-// Reference: https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/system/freertos.html
-
 use anyhow::Result;
-use esp_idf_svc::sys::{
-    esp, gpio_config, gpio_config_t, gpio_install_isr_service, gpio_int_type_t_GPIO_INTR_POSEDGE,
-    gpio_isr_handler_add, gpio_mode_t_GPIO_MODE_INPUT, xQueueGenericCreate, xQueueGiveFromISR,
-    xQueueReceive, QueueHandle_t, ESP_INTR_FLAG_IRAM,
+use esp_idf_svc::hal::{
+    gpio::{InterruptType, PinDriver, Pull},
+    peripherals::Peripherals,
+    task::notification::Notification,
 };
-use std::ptr;
-
-// This `static mut` holds the queue handle we are going to get from `xQueueGenericCreate`.
-// This is unsafe, but we are careful not to enable our GPIO interrupt handler until after this value has been initialised, and then never modify it again
-static mut EVENT_QUEUE: Option<QueueHandle_t> = None;
-
-#[link_section = ".iram0.text"]
-unsafe extern "C" fn button_interrupt(_: *mut core::ffi::c_void) {
-    xQueueGiveFromISR(EVENT_QUEUE.unwrap(), std::ptr::null_mut());
-}
+use std::num::NonZeroU32;
 
 fn main() -> Result<()> {
-    const GPIO_NUM: i32 = 9;
+    esp_idf_svc::sys::link_patches();
+
+    let peripherals = Peripherals::take()?;
 
     // Configures the button
-    let io_conf = gpio_config_t {
-        pin_bit_mask: 1 << GPIO_NUM,
-        mode: gpio_mode_t_GPIO_MODE_INPUT,
-        pull_up_en: true.into(),
-        pull_down_en: false.into(),
-        intr_type: gpio_int_type_t_GPIO_INTR_POSEDGE, // Positive edge trigger = button down
-    };
+    let mut button = PinDriver::input(peripherals.pins.gpio9)?;
+    button.set_pull(Pull::Up)?;
+    button.set_interrupt_type(InterruptType::PosEdge)?;
 
-    // Queue configurations
-    const QUEUE_TYPE_BASE: u8 = 0;
-    const ITEM_SIZE: u32 = 0; // We're not posting any actual data, just notifying
-    const QUEUE_SIZE: u32 = 1;
+    // Configures the notification
+    let notification = Notification::new();
+    let notifier = notification.notifier();
 
+    // Safety: make sure the `Notification` object is not dropped while the subscription is active
```

Comment:
also make sure because this is the training to educate people about, that the closure is running in the ISR itself, so to stop them making the mistake that everybody does and call prinltn from it. So no standard library calls inside it - and more generally nothing that blocks in any way